sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -146,21 +159,119 @@ public String getMimeType() {
    */
   protected final WritableByteChannelFactory writableByteChannelFactory;
 
+
   /**
-   * Base filename for final output files.
+   * A naming policy for output files.
    */
-  protected final ValueProvider<String> baseOutputFilename;
+  public abstract static class FilenamePolicy
+      implements org.apache.beam.sdk.transforms.SerializableFunction<
+       FilenamePolicy.Context, String> {
+    /**
+     * Context used for generating a name based on window, pane, shard numer, and num shards.
+     * Window and pane will only be provided if windowed writes have been requested using
+     * {@link Write.Bound#withWindowedWrites}. The policy must produce unique filenames for unique
+     * Context objects.
+     */
+    public static class Context {
+      @Nullable private BoundedWindow window;
+      @Nullable private PaneInfo paneInfo;
+      private int shardNumber;
+      private int numShards;
+
+      public Context(
```

Comment:
After considering this, I think this API is a bit dangerous. If these billable parameters are provided, and the policy ignores them, it likely constitutes data loss. Really, they should be all non-nullable and the old API (if we keep it at all) should be a separate method. A sink that does not have the windowed method should be statically rejected for any non-default windowing or triggering.